### PR TITLE
Upgrade to Selenium 4

### PR DESCRIPTION
**Pull request by jtnord**
Based on #970 at least some tests work on my machine...

@Vlatombe 

- [x] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [ ] Please describe what you did
- [ ] Link to relevant issues in GitHub or Jira
- [ ] Link to relevant pull requests, esp. upstream and downstream changes
- [x] Ensure you have provided tests - that demonstrates feature works or fixes the issue

### CODE REVIEW

File: src/main/resources/ath-container/Dockerfile [hunk 1]
```diff
@@ -22,8 +22,8 @@ RUN yum -y install \
     yum clean all && rm -rf /var/cache/yum
 
 # Selenium needs a geckodriver in order to work properly
-RUN curl -fsSLO https://github.com/mozilla/geckodriver/releases/download/v0.26.0/geckodriver-v0.26.0-linux64.tar.gz && \
-    tar -xvzf geckodriver-v0.26.0-linux64.tar.gz -C /usr/local/bin
+RUN curl -fsSLO https://github.com/mozilla/geckodriver/releases/download/v0.32.0/geckodriver-v0.32.0-linux-aarch64.tar.gz && \
```
[the remark this answers is not in the document]
Comment (timja) [reply src/main/resources/ath-container/Dockerfile:25]: are you intentionally switching this to arm?

there's some code in this PR which handles different architectures properly (hopefully) https://github.com/jenkinsci/acceptance-test-harness/pull/878

it should be simpler now that geckodriver publishes arm builds

[the remark this answers is not in the document]
Comment (jtnord) [reply src/main/resources/ath-container/Dockerfile:25]: oops.  no I was not - yet all bar one of the tests pass :-o

[the remark this answers is not in the document]
Comment (jtnord) [reply src/main/resources/ath-container/Dockerfile:25]:
```suggestion
RUN curl -fsSLO https://github.com/mozilla/geckodriver/releases/download/v0.32.0/geckodriver-v0.32.0-linux64.tar.gz && \
```

**Comment by jtnord** [comment]
I consider this ready for review but please do not merge.  I have some internal test suites to update that may changes and they need to be done simultaneously :-(

**Comment by jtnord** [comment]
running locally 

```
[ERROR]   ConfigFileProviderTest.testCustomConfigFile:99->createCustomConfig:110->CapybaraPortingLayerImpl.clickButton:100 » ElementNotInteractable Element <button id="yui-gen2-button" type="submit"> could not be scrolled into view
Build info: version: '4.5.0', revision: 'fe167b119a'
System info: os.name: 'Windows 11', os.arch: 'amd64', os.version: '10.0', java.version: '11.0.15'
Driver info: org.openqa.selenium.remote.RemoteWebDriver
Command: [dbf85f5e-76cb-44a2-8174-bec582c58118, clickElement {id=17afee69-5ae6-4ea4-8fc9-d5521037a224}]
Capabilities {acceptInsecureCerts: true, browserName: firefox, browserVersion: 105.0.2, moz:accessibilityChecks: false, moz:buildID: 20221003135607, moz:debuggerAddress: 127.0.0.1:46002, moz:firefoxOptions: {prefs: {dom.disable_beforeunload: false, dom.max_chrome_script_run_time: 600000, dom.max_script_run_time: 600000, intl.accept_languages: en}}, moz:geckodriverVersion: 0.31.0, moz:headless: false, moz:platformVersion: 5.10.102.1-microsoft-standa..., moz:processID: 1233, moz:profile: /tmp/rust_mozprofile3LR86T, moz:shutdownTimeout: 60000, moz:useNonSpecCompliantPointerOrigin: false, moz:webdriverClick: true, moz:windowless: false, pageLoadStrategy: normal, platformName: LINUX, proxy: Proxy(manual, http=172.30.2..., se:cdp: ws://172.17.0.2:4444/sessio..., se:cdpVersion: 85.0, se:noVncPort: 7900, se:vnc: ws://172.17.0.2:4444/sessio..., se:vncEnabled: true, se:vncLocalAddress: ws://172.17.0.2:7900, setWindowRect: true, strictFileInteractability: false, timeouts: {implicit: 0, pageLoad: 300000, script: 30000}, unhandledPromptBehavior: dismiss and notify}
Element: [[RemoteWebDriver: firefox on LINUX (dbf85f5e-76cb-44a2-8174-bec582c58118)] -> xpath: .//input[./@type = 'submit' or ./@type = 'reset' or ./@type = 'image' or ./@type = 'button'][((./@id = 'Submit' or ./@name = 'Submit' or contains(./@value, 'Submit')) or contains(./@title, 'Submit'))] | .//input[./@type = 'image'][contains(./@alt, 'Submit')] | .//button[(((./@id = 'Submit' or contains(./@value, 'Submit')) or contains(normalize-space(.), 'Submit')) or contains(./@title, 'Submit'))] | .//input[./@type = 'image'][contains(./@alt, 'Submit')]]
Session ID: dbf85f5e-76cb-44a2-8174-bec582c58118
```

using `docker run --shm-size=256m -d -p 127.0.0.1:4444:4444 -p 127.0.0.1:5900:5900 -e no_proxy=localhost -e SCREEN_WIDTH=1680 -e SCREEN_HEIGHT=1090 selenium/standalone-firefox:4.5.0`

I see the same failure when running in our CI environment :-/

File: vars.cmd [hunk 2]
```diff
@@ -23,4 +23,4 @@ set SELENIUM_PROXY_HOSTNAME=%IP%
 set JENKINS_LOCAL_HOSTNAME=%IP%
 @echo.
 @echo To start the remote firefox container run the following command:
-@echo docker run --shm-size=256m -d -p 127.0.0.1:4444:4444 -p 127.0.0.1:5900:5900 -e no_proxy=localhost -e SCREEN_WIDTH=1680 -e SCREEN_HEIGHT=1090 selenium/standalone-firefox-debug:3.141.59
+@echo docker run --shm-size=256m -d -p 127.0.0.1:4444:4444 -p 127.0.0.1:5900:5900 -e no_proxy=localhost -e SCREEN_WIDTH=1680 -e SCREEN_HEIGHT=1090 selenium/standalone-firefox:4.5.0
```
Comment (jtnord) [inline vars.cmd:26]:
```suggestion
@echo docker run --shm-size=256m -d -p 127.0.0.1:4444:4444 -p 127.0.0.1:5900:5900 -e no_proxy=localhost -e SCREEN_WIDTH=1680 -e SCREEN_HEIGHT=1090 selenium/standalone-firefox:4.6.0

```

File: pom.xml [hunk 3]
```diff
@@ -31,7 +31,7 @@
     <trimStackTrace>false</trimStackTrace> <!-- surefire -->
     <spotbugs.skip>true</spotbugs.skip>
 
-    <selenium.version>3.141.59</selenium.version>
+    <selenium.version>4.5.0</selenium.version>
```
Comment (jtnord) [inline pom.xml:34]:
```suggestion
    <selenium.version>4.6.0</selenium.version>
```

4.5.0 had an important[ thread concurrency issue that caused random flakes](https://github.com/SeleniumHQ/docker-selenium/issues/1507) that we need to pickup.

File: pom.xml [hunk 4]
```diff
@@ -443,6 +431,25 @@
         <artifactId>commons-io</artifactId>
         <version>2.11.0</version>
       </dependency>
+      <dependency>
+        <groupId>org.apache.commons</groupId>
+        <artifactId>commons-text</artifactId>
+        <version>1.10.0</version>
+        <!-- 
+        commons confif is only used by org/jenkinsci/test/acceptance/utils/pluginreporter/TextFileExercisedPluginReporter
```
Comment (timja) [inline pom.xml:439]:
```suggestion
        commons config is only used by org/jenkinsci/test/acceptance/utils/pluginreporter/TextFileExercisedPluginReporter
```

**Comment by jtnord** [comment]
our XPATH is too big...

```
Caused by: javax.xml.xpath.XPathExpressionException: javax.xml.transform.TransformerException: JAXP0801002: the compiler encountered an XPath expression containing '101' operators that exceeds the '100' limit set by 'FEATURE_SECURE_PROCESSING'.
Caused by: javax.xml.transform.TransformerException: JAXP0801002: the compiler encountered an XPath expression containing '101' operators that exceeds the '100' limit set by 'FEATURE_SECURE_PROCESSING'.

**Comment by jtnord** [comment]
3 failures are all around agent configuration and `Element <button class="jenkins-button jenkins-button--primary " name="Submit"> is not clickable at point (1535,268) because another element <div class="timestamp"> obscures it`

Likely the scrollable hook is not being used here (or something else) - would be nice to have a screenshot.

**Comment by timja** [comment]
> 3 failures are all around agent configuration and `Element <button class="jenkins-button jenkins-button--primary " name="Submit"> is not clickable at point (1535,268) because another element <div class="timestamp"> obscures it`

SlaveTest is probably https://github.com/jenkinsci/acceptance-test-harness/pull/1018#issuecomment-1384637525

**Comment by jtnord** [comment]
> > 3 failures are all around agent configuration and `Element <button class="jenkins-button jenkins-button--primary " name="Submit"> is not clickable at point (1535,268) because another element <div class="timestamp"> obscures it`
> 
> SlaveTest is probably [#1018 (comment)](https://github.com/jenkinsci/acceptance-test-harness/pull/1018#issuecomment-1384637525)

Thanks Tim.  Looks like the same 3 failures.

File: src/main/java/org/jenkinsci/test/acceptance/selenium/Scroller.java [hunk 5]
```diff
@@ -74,7 +74,7 @@
  * @author ogondza
  * @author Kohsuke Kawaguchi
  */
-public class Scroller extends AbstractWebDriverEventListener {
+public class Scroller extends  AbstractWebDriverEventListener {
```
[the remark this answers is not in the document]
Comment (timja) [reply src/main/java/org/jenkinsci/test/acceptance/selenium/Scroller.java:77]:
```suggestion
public class Scroller extends AbstractWebDriverEventListener {
```

File: vars.cmd [hunk 6]
```diff
@@ -23,4 +23,5 @@ set SELENIUM_PROXY_HOSTNAME=%IP%
 set JENKINS_LOCAL_HOSTNAME=%IP%
 @echo.
 @echo To start the remote firefox container run the following command:
-@echo docker run --shm-size=256m -d -p 127.0.0.1:4444:4444 -p 127.0.0.1:5900:5900 -e no_proxy=localhost -e SCREEN_WIDTH=1680 -e SCREEN_HEIGHT=1090 selenium/standalone-firefox-debug:3.141.59
+@echo docker run --shm-size=256m -d -p 127.0.0.1:4444:4444 -p 127.0.0.1:5900:5900 -e no_proxy=localhost -e SCREEN_WIDTH=1680 -e SCREEN_HEIGHT=1090 selenium/standalone-firefox:4.6.0
+
```
Comment (timja) [inline vars.cmd:27]:
```suggestion

```

**Review by timja — COMMENTED** [review]
All passing now, I pushed [`70a2ea5` (#973)](https://github.com/jenkinsci/acceptance-test-harness/pull/973/commits/70a2ea51851c630e55d458e74dbf2568459c3efc) to work around a known issue with the reverse proxy on updates.jenkins.io.

good to go @jtnord?

File: vars.cmd (hunk 6 above)
Comment (jtnord) [reply vars.cmd:27]: trailing newline because UNIX tools (even though this is a windows thing...)  messes with `cat` when I use it :)

**Review by raul-arabaolaza — APPROVED** [review]
Code looks good to me, I will check the test failures to see what is happening

**Comment by raul-arabaolaza** [comment]
That installWizardTest is flaky, it has happened to fail to me in other builds, I will try to try to unflake it here or create a ticket and ignore if I am not able to do it

**Comment by raul-arabaolaza** [comment]
Yep, definitively flaky, it looks like the continue button in the wizard is not clicked, or because there is no wait after the click and the wizard does not load the next page fast enough sometimes

**Comment by raul-arabaolaza** [comment]
Well, my initial tought is not correct, as after trying to click the button the next find waits 120 seconds for the finish button to appear...

So somehow the button to confirm the url settings should not be clicked or the click not work sometimes...

**Comment by uhafner** [comment]
This breaks all of my tests:
https://github.com/jenkinsci/warnings-ng-plugin/pull/1470

Is there an upgrade guide available?

**Comment by timja** [comment]
It appears to be failing for missing netty
`Caused by: java.lang.NoClassDefFoundError: io/netty/handler/codec/DefaultHeaders$ValueValidator`

odd it should be coming transitively...

**Comment by raul-arabaolaza** [comment]
@uhafner I suspect you may be impacted by https://issues.apache.org/jira/browse/MNG-5761 we were impacted on our private tests suites also. So we had to add the netty bom in your dependencyManagemnt to circumvent the maven issue which was preventing the proper netty codec version to be resolved transitively.  https://github.com/jenkinsci/acceptance-test-harness/blob/master/pom.xml#L402-L409

**Comment by uhafner** [comment]
@raul-arabaolaza Thanks, that fixed the problem for the warnings plugin as well (see https://github.com/jenkinsci/warnings-ng-plugin/pull/1470/commits/c3a4ae26a575fb62f22e43591fa8adff33433499)! Finally, all my UI tests run with latest ATH and latest Jenkins again. 

🚀

**Pull request merged**